### PR TITLE
Add reduce scripts

### PR DESCRIPTION
**Pull request by JustinStitt**
> **Note**
> This PR had a previous home at https://github.com/ClangBuiltLinux/misc-scripts/pull/4
> But now lives here as suggested by: https://github.com/ClangBuiltLinux/misc-scripts/pull/4#pullrequestreview-1546542798

This PR aims to introduce a new family of scripts aimed at helping kernel developers reduce target files efficiently.

An example workflow may look something like this:
```
> CI robot identifies build failure in some kernel source tree

> Developer wants to isolate the issue and find its root cause

> Developer builds that source tree version and, sure enough, replicates the issue. They're still lost as to what is causing the failure.

> Developer runs `$ python reduce.py prep lib/broken`

> Developer now has access to `broken.i`, `test.sh` and `flags.txt`.

> Developer designs interestingness test that fully encompasses the issue/behavior they're looking for.

> Developer wants to minimize the number of compiler flags being used and thus runs `$ python reduce.py flags`

> Now, there is a minimized `flags.txt`.

> They then run `$ cvise test.sh broken.i` (and wait awhile).

The developer now has both 1) the minimal flags and 2) minimal pre-processed source to reproduce some behavior
```

Another possible feature to come is one that automatically creates a `repro.txt` file which adequately displays the necessary information to reproduce build failures (i.e: compiler version, minimal flags used, minimal pre-processed code, kernel version, etc). The idea is that this `repro.txt` can then be shared with others (via mailing lists or chats).

It should be noted these scripts are **not** zero-dependency. Their dependencies are listed in the README.

Below is a copy of the README.md.

# CBL/reduce
A family of scripts to prepare and execute reduction (cvise, llvm-reduce) workflows.

Currently, there are two scripts:

1) prep
> Prepare a target file for reduction (see Basic Usage)

2) flags
> Minimize the compiler flags required to reproduce behavior. Use after prep

## Dependencies
* Python 3.11
* [icecream](https://pypi.org/project/icecream/)
* [C-Vise](https://github.com/marxin/cvise)

## Basic Usage

`$ python reduce.py -h`

`$ python reduce.py prep -h`

> Prepare lib/string for reduction by producing a
> `string.i`, `test.sh` and a `flags.txt` in a temporary directory which are
> ready-for-consumption by tools like `cvise` and `llvm-reduce`

`$ python reduce.py prep lib/string -p /path/to/linux -o $(mktemp -d)`

> Or, try providing a custom build command
> 
`$ reduce prep -p $LINUX -o $(mktemp -d) -- make -j8 LLVM=1 V=1 lib/string`

`$ python reduce.py flags -h`

> After running `prep`, reduce flags.txt

`$ python reduce.py flags`

> **Note**
> Change your cwd to wherever `flags.txt` is located or try the `-p` option.

## Contributions
Author: Justin Stitt <justinstitt@google.com>

Suggested-by: Nick Desaulniers <ndesaulniers@google.com>

### CODE REVIEW

**Review by nathanchance — APPROVED** [review]
Thanks, still LGTM, although now that this is in its own repo, I think we can move everything into the repo root, rather than having everything under `reduce/`?

Consider adding a [`CODEOWNERS` file](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) like we have in other repositories to ensure someone always gets added for review and standing up some linting/formatting workflows to ensure that the code stays consistent. Follow up PRs are fine for those, they are not mission critical.

**Comment by JustinStitt** [comment]
> Thanks, still LGTM, although now that this is in its own repo, I think we can move everything into the repo root, rather than having everything under `reduce/`?
> 
> Consider adding a [`CODEOWNERS` file](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners) like we have in other repositories to ensure someone always gets added for review and standing up some linting/formatting workflows to ensure that the code stays consistent. Follow up PRs are fine for those, they are not mission critical.

Yep, just did this before seeing your question. My copying from repo to repo resulted in a nested folder. Fixed. Thanks.

**Comment by JustinStitt** [comment]
I will leave the PR up for code reviewing purposes. Let me know once its OK to merge into CBL/reduce.

File: README.md
Comment (nickdesaulniers) [inline README.md:34]: should this be

$ python reduce.py prep ....

File: README.md
Comment (nickdesaulniers) [inline README.md:55]: You can remove these.  Folks can use github UI or git log if they care for stats.

File: src/cli.py
Comment (nickdesaulniers) [inline src/cli.py:52]: Is there a more specific exception type you can catch here?

File: src/cli.py
Comment (nickdesaulniers) [inline src/cli.py:1]:
```python
from argparse import ArgumentTypeError
```

File: src/cli.py
Comment (nickdesaulniers) [inline src/cli.py:92]: Will `".i"` scan the entire string? Maybe check the string ends with ".i"?

File: src/cli.py
Comment (nickdesaulniers) [inline src/cli.py:90]: Is this requiring the `make` invocation to contain `V=1`? Seems like we should just silently add that, if so.

File: src/prep.py
Comment (nickdesaulniers) [inline src/prep.py:262]: an optional bool? is that really necessary?

File: src/prep.py
Comment (nickdesaulniers) [inline src/prep.py:95]: should `rule not in flag` instead be `not flag.startswith(rule)` rather than scan the entire flag?

File: src/prep.py
Comment (nickdesaulniers) [inline src/prep.py:109]: what does `*` in a parameter list mean?

File: src/prep.py
Comment (nickdesaulniers) [inline src/prep.py:129]: what is this slicing trying to do? Not clear from the comments

File: src/prep.py
Comment (nickdesaulniers) [inline src/prep.py:133]: what is this slicing trying to do? Not clear from the comments

File: README.md
Comment (nickdesaulniers) [inline README.md:24]: needs more Preparation-H jokes

File: src/cli.py
Comment (JustinStitt) [reply src/cli.py:52]: I've opted to remove this try/except completely.

Nathan mentioned this earlier too: https://github.com/ClangBuiltLinux/misc-scripts/pull/2#discussion_r1269701328

It is, in fact, pretty impossible to conjure up a string that pathlib can't convert to a path.

File: src/prep.py
Comment (JustinStitt) [reply src/prep.py:95]: `in` should stop once it finds it (presumably at the beginning) and not proceed to scan the whole string.

I think `in` is better than `startswith` because I believe I've seen a flag like `"./lib/..."` (with quotes present) which means `./` wouldn't be the start of the flag.

Can you confirm if you've seen a flag wrapped in quotes like that or perhaps it was some part of my processing of the flags that added a quote and that's what I'm remembering.

File: src/prep.py
Comment (JustinStitt) [reply src/prep.py:109]: Mandatory kwargs to follow.

```py
def foo(*, a: int, b: int) -> int:
   ...

# now any invocation of foo requires kwargs
result = foo(a=4, b=6)
```

Comment (JustinStitt) [reply src/prep.py:109]: something like 
```py
foo(3, 4) # raises TypeError: foo() takes 0 positional arguments but 2 were given
```

Comment (JustinStitt) [reply src/prep.py:109]: I usually add `(*, ...)` if my function takes 3 or 4 very similar argumemts (like paths, or flags, or numbers) so that the caller cannot mess it up. (also helps my linter give me argument-hints when I'm typing an invocation)

File: src/prep.py
Comment (JustinStitt) [reply src/prep.py:133]: turn

`clang -Wall -Wsomething -Wotherthing -xyz --target=thermometer -o target.out -c target.i`

into `-Wall -Wsomething -Wotherthing -xyz --target=thermometer`

essentially ditching the `clang` and `-o target.out -c target.i` so I can write them to a flags.txt file. Our flags.txt file need not modify our compiler invocation nor our target i/o.

Comment (JustinStitt) [reply src/prep.py:133]: Any ideas on a more elegant approach than magic-number slicing?

File: src/prep.py
Comment (nickdesaulniers) [reply src/prep.py:109]: neat! TIL

File: README.md
Comment (JustinStitt) [reply README.md:24]: gpt was glad to help.

```
Sure, here are some humorous lines you might find amusing:                                                                                                                                                                                      ▎   ▎ 
 ▊ ▊                                                                                                                                                                                                                                                    ▎   ▎ 
 ▊ ▊      1. "No side effects guaranteed, only code effects!"                                                                                                                                                                                           ▎   ▎ 
 ▊ ▊      2. "Coding with 'prep -h' may relieve symptoms of buggy code, unlike Preparation-H!"                                                                                                                                                          ▎   ▎ 
 ▊ ▊      3. "Warning: Excessive use of 'prep -h' may result in becoming a Python expert, not a hemorrhoid treatment expert!"                                                                                                                           ▎   ▎ 
 ▊ ▊      4. "Relieves code constipation. More effective than Preparation-H!"                                                                                                                                                                           ▎   ▎ 
 ▊ ▊      5. "For highly irritating bugs, we recommend 'prep -h', but for everything else, there's Preparation-H."                                                                                                                                      ▎   ▎ 
 ▊ ▊      6. "Helps your code get back on track. Soothing as Preparation-H but strictly for programmers!"                                                                                                                                               ▎   ▎ 
 ▊ ▊      7. "Unlike Preparation-H, 'prep -h' doesn't require application once every 4 hours!"                                                                                                                                                          ▎   ▎ 
 ▊ ▊      8. "'prep -h' - the only relief you need from memory leaks, not other leaks!"                                                                                                                                                                 ▎   ▎ 
 ▊ ▊      9. "'prep -h', the only ointment your code needs!"                                                                                                                                                                                            ▎   ▎ 
 ▊ ▊     10. "Side effects may include cleaner code and an uncontrollable urge to program! Far better than Preparation-H."
```

Comment (JustinStitt) [reply README.md:24]: Personal favorite: 

"Unlike Preparation-H, 'prep -h' doesn't require application once every 4 hours!"

File: README.md
Comment (JustinStitt) [reply README.md:34]: fixed. thanks.

File: src/prep.py
Comment (nickdesaulniers) [reply src/prep.py:95]: The problem I have with `in` is when it doesn't find the needle in the haystack.  startswith will not look further into the haystack then the length of the needle.

File: src/cli.py
Comment (JustinStitt) [reply src/cli.py:1]: Great suggestion. fixed.

File: src/cli.py
Comment (JustinStitt) [reply src/cli.py:92]: Got it. fixed.

File: src/prep.py
Comment (JustinStitt) [reply src/prep.py:262]: Probably not. Fixed.

File: src/cli.py
Comment (JustinStitt) [reply src/cli.py:90]: Took your suggestion, it's silently (with an info message) added now.

File: src/prep.py
Comment (JustinStitt) [reply src/prep.py:129]: duplicate comment, closing.

File: src/types.py
Comment (nickdesaulniers) [inline src/types.py:3]: The `| Path` part seems to lead to casts to `str` in at least `write_test_script`. Would it be simpler to have this be a list of `str`, delete this file, then have casts to `Path` when necessary?

File: src/prep.py
Comment (nickdesaulniers) [reply src/prep.py:133]: Assuming that `-o foo.o` and `-c input.i` are the last 4 positional command line args is brittle.  It's probably the case for all code in the kernel today, but I don't know that is true or always will be.

Explicitly filtering out `-o <and what follows -o>` is more self documenting than slicing by magic values.  Same for `-c` and the input filename.

This is not the case with the initial arg, since `clang` can't appear like `-Wall -Wsomething clang -Wotherthing -xyz --target=thermometer -o target.out -c target.i`

Comment (JustinStitt) [reply src/prep.py]: Fixed. Thanks.

File: src/prep.py
Comment (nickdesaulniers) [inline src/prep.py:143]: careful, target.c doesn't necessarily follow `-c` like `target.o` follows `-o`. (`-c` is position independent relative to other args).

Do we have the target name say from perhaps the caller of `cc_invocation_to_flags`?

File: src/prep.py
Comment (nickdesaulniers) [reply src/prep.py:116]: > I think in is better than startswith because I believe I've seen a flag like "./lib/..." (with quotes present) which means ./ wouldn't be the start of the flag.

Ah, sorry I missed your point about this.

Do command line args beginning with `./` exist BOTH quoted and unquoted?

Taking a step back, `./some/path` doesn't make sense to remove in isolation.

Some flags are pairs, like `-o` and whatever follows it.  In such cases we should be removing the flag AND whatever follows it.

As such, I suspect that flags like `./some/path` (regardless of whether they are quoted or not) aren't really position independent flags like `-c` is.  As such, removing them one at a time might miss removing the flag that comes before them.

In this case I suspect that `./some/path` flags follow `-I` or `-include`.  As such, we should treat those (and `-D` and `-U`) with more care.  We need to handle those differently.  In particular, `-I`, `-D`, and `-U` may or may not be space delimated from their arguments.

`-Dfoo` is equivalent to `-D foo`. So if we split on spaces, we need to check if the flag that starts with `-D` has a space immediately after the `D`; if so remove `-D` and the following flag, if not then remove the entire flag.  Same for the above.

Of the current flags in `to_remove`, it only makes sense to remove `-E` and `-Werror` one by one, as they don't imply any positional relationship to other flags.

Comment (JustinStitt) [reply src/prep.py]: Got it, this requires a semi-substantial refactor of the current flag-cleaning processes.

Will think it out.

File: src/prep.py
Comment (JustinStitt) [reply src/prep.py:145]: Whoops, I naively assumed `-c` was positional.

Will fix.

File: src/types.py
Comment (JustinStitt) [reply src/types.py:3]: Going to keep as is. This is used for type hinting only and makes my linter not hate me.

Some of the casts from `Path -> str` can be done with `my_path.name` and I will look to see if any code can be made cleaner this way.

**Pull request merged**
